### ISSUE TITLE
Isn't the weight learnt from each minibatchs are supposed to be averaged?

### ISSUE THREAD

**Issue by feather0011** [timgaripov/swa#18]
Hello, thank you for the great work.

I am trying to use swa in my work, so I read code to find out what I get from the paper is right.

However, I think that the way the weight is averaged is different.

In paper( and also in [other_example](https://github.com/izmailovpavel/contrib_swa_examples/blob/master/train.py)), it seems that weights learnt from each minibatch are ensembled at the end of an epoch. 

However, in this code, it seems that weights learnt form each epoch are ensembled.

So I am confused.

**Comment by izmailovpavel** [comment]
Hey @feather0011, we collect and average weights once per epoch in this code and in the paper. In the `torch.contrib` implementation you can specify frequency of averaging. In general, SWA works well if you average more often than once per epoch, even after each mini-batch. We only average once per epoch for convenience and to avoid overhead from averaging too often.

**Comment by feather0011** [comment]
Thank you for the quick answer!

However, I think the answer is not about what I had pointed.

In [torchcontrib.optim.swa](https://github.com/pytorch/contrib/blob/master/torchcontrib/optim/swa.py), [update_swa_group](https://github.com/pytorch/contrib/blob/c545fedf4f73c8e95f91fd81f2d5bf7fa9c62a61/torchcontrib/optim/swa.py#L150), which adds current weights to "param_state['swa_buffer']", are called each step of optimization.

This mean that weights trained by **each minibatch** are averaged once per epoch.

I think this is same to what the paper described in page 4 : _"Here by iteration we assume the processing of one batch of data"_ 

However, in the code, the weights those are collected and averaged are **weights from each previous epoch**. 

[This is the code part of train.py](https://github.com/timgaripov/swa/blob/f54af44f7af3cadd628412bbc34383e94e56ae53/train.py#L154)

`
for epoch in range(start_epoch, args.epochs):
    time_ep = time.time()
   
    lr = schedule(epoch)
    utils.adjust_learning_rate(optimizer, lr)
    train_res = utils.train_epoch(loaders['train'], model, criterion, optimizer)
    if epoch == 0 or epoch % args.eval_freq == args.eval_freq - 1 or epoch == args.epochs - 1:
        test_res = utils.eval(loaders['test'], model, criterion)
    else:
        test_res = {'loss': None, 'accuracy': None}
    
    if args.swa and (epoch + 1) >= args.swa_start and (epoch + 1 - args.swa_start) % args.swa_c_epochs == 0:
        utils.moving_average(swa_model, model, 1.0 / (swa_n + 1))
        swa_n += 1
        if epoch == 0 or epoch % args.eval_freq == args.eval_freq - 1 or epoch == args.epochs - 1:
            utils.bn_update(loaders['train'], swa_model)
            swa_res = utils.eval(loaders['test'], swa_model, criterion)
        else:
            swa_res = {'loss': None, 'accuracy': None}
`

Here, weights which are collected from end of each epochs are avergaed.

The differene I want to point out can be easily seen when one compares how "swa_n" and "group["n_avg"]" in each code are updated.

According to my understanding, both implies the number of weight to be collected and averaged. 

In [train.py](https://github.com/timgaripov/swa/blob/f54af44f7af3cadd628412bbc34383e94e56ae53/train.py#L154) "swa_n" is updated only once per epoch. 

However, in [torchcontrib.optim.swa](https://github.com/pytorch/contrib/blob/master/torchcontrib/optim/swa.py),  "group["n_avg"]" is updated when the swa.step() is called, meaning it increases N times every epoch if one epoch is composed of N minibatchs. 

P.S Sorry for messing up the code block. I could not figure out how to use it properly. It must include 
'for epoch in range(start_epoch, args.epochs):
    time_ep = time.time()'

**Comment by izmailovpavel** [comment]
Hey @feather0011, I am not sure I understand what you mean by 
>  This mean that weights trained by each minibatch are averaged once per epoch.

In SWA we train the network as usual, and average the weights once per epoch. In `torchcontrib` the weights are averaged once every `swa_freq` steps: https://github.com/pytorch/contrib/blob/master/torchcontrib/optim/swa.py#L209-L222 
The idea is to set `swa_freq` to the number of steps (mini-batches) in one epoch.

In particular, notice this part of the `step()` function on `torchcontrib`
```
if steps > self.swa_start and steps % self.swa_freq == 0:
    self.update_swa_group(group)
```

You are correct that in the code in this repo we average the weights once per epoch.

**Comment by feather0011** [comment]
Dear, @izmailovpavel. I wrote down pseudo code to represent how I think two codes are different.

I assumed 'swa_freq = 1'

In this repo  :
```
swa_buffer = zeros_like(model_weight)
swa_n = 0
for epoch_idx in range(swa_start_epoch, total_epoch):
    swa_n+=1
    for batch_idx in range(batch_per_epoch):
        model_weight = SGD(model_weight)
    swa_buffer = ((n-1)/n)*swa_buffer +  (1/n)*model_weight   # weights from the ends of previous epochs are averaged.
    model_weight = swa_buffer      # update (once per epoch)
```
 
in `torchcontrib` :
```
swa_buffer = zeros_like(model_weight)
swa_n = 0
for epoch_idx in range(swa_start_epoch, total_epoch):
    for batch_idx in range(batch_per_epoch):
        swa_n+=1
        model_weight = SGD(model_weight)
        swa_buffer = ((n-1)/n)*swa_buffer +  (1/n)*model_weight   # weights from previous minibatchs are averaged.
    model_weight = swa_buffer      # update (once per epoch)
```

In `torchcontrib`, [`step`](https://github.com/pytorch/contrib/blob/c545fedf4f73c8e95f91fd81f2d5bf7fa9c62a61/torchcontrib/optim/swa.py#L209), which is expected to be called every minibatch, calls `update_swa_group`. 

For my own research, averaging like the second one showed more stable result.
(The best performance was similar when performing early stopping though.)

**Comment by izmailovpavel** [comment]
Hey @feather0011, I think the reason the two examples are different is this assumption:
> I assumed `swa_freq = 1`

The two repos are equivalent if you set `swa_freq` to the number of mini-batches in one epoch. However, as long as `swa_freq=1` works better for you, you should use it! Is there any reason you need to use this repo over `torchcontrib` implementation? 

Btw, there is a native implementation of SWA coming in PyTorch 1.6 (https://github.com/pytorch/pytorch/blob/master/torch/optim/swa_utils.py), which should be released very soon :) You will be able to manually decide on the frequency of SWA updates there, like in `torchcontrib`.
